plugins/inputs/riak/riak.go

```diff
@@ -137,51 +190,104 @@ func (r *Riak) gatherServer(s string, acc telegraf.Accumulator) error {
 
 	// Build a map of field values
 	fields := map[string]interface{}{
-		"cpu_avg1":                     stats.CPUAvg1,
-		"cpu_avg15":                    stats.CPUAvg15,
-		"cpu_avg5":                     stats.CPUAvg5,
-		"memory_code":                  stats.MemoryCode,
-		"memory_ets":                   stats.MemoryEts,
-		"memory_processes":             stats.MemoryProcesses,
-		"memory_system":                stats.MemorySystem,
-		"memory_total":                 stats.MemoryTotal,
-		"node_get_fsm_objsize_100":     stats.NodeGetFsmObjsize100,
-		"node_get_fsm_objsize_95":      stats.NodeGetFsmObjsize95,
-		"node_get_fsm_objsize_99":      stats.NodeGetFsmObjsize99,
-		"node_get_fsm_objsize_mean":    stats.NodeGetFsmObjsizeMean,
-		"node_get_fsm_objsize_median":  stats.NodeGetFsmObjsizeMedian,
-		"node_get_fsm_siblings_100":    stats.NodeGetFsmSiblings100,
-		"node_get_fsm_siblings_95":     stats.NodeGetFsmSiblings95,
-		"node_get_fsm_siblings_99":     stats.NodeGetFsmSiblings99,
-		"node_get_fsm_siblings_mean":   stats.NodeGetFsmSiblingsMean,
-		"node_get_fsm_siblings_median": stats.NodeGetFsmSiblingsMedian,
-		"node_get_fsm_time_100":        stats.NodeGetFsmTime100,
-		"node_get_fsm_time_95":         stats.NodeGetFsmTime95,
-		"node_get_fsm_time_99":         stats.NodeGetFsmTime99,
-		"node_get_fsm_time_mean":       stats.NodeGetFsmTimeMean,
-		"node_get_fsm_time_median":     stats.NodeGetFsmTimeMedian,
-		"node_gets":                    stats.NodeGets,
-		"node_gets_total":              stats.NodeGetsTotal,
-		"node_put_fsm_time_100":        stats.NodePutFsmTime100,
-		"node_put_fsm_time_95":         stats.NodePutFsmTime95,
-		"node_put_fsm_time_99":         stats.NodePutFsmTime99,
-		"node_put_fsm_time_mean":       stats.NodePutFsmTimeMean,
-		"node_put_fsm_time_median":     stats.NodePutFsmTimeMedian,
-		"node_puts":                    stats.NodePuts,
-		"node_puts_total":              stats.NodePutsTotal,
-		"pbc_active":                   stats.PbcActive,
-		"pbc_connects":                 stats.PbcConnects,
-		"pbc_connects_total":           stats.PbcConnectsTotal,
-		"vnode_gets":                   stats.VnodeGets,
-		"vnode_gets_total":             stats.VnodeGetsTotal,
-		"vnode_index_reads":            stats.VnodeIndexReads,
-		"vnode_index_reads_total":      stats.VnodeIndexReadsTotal,
-		"vnode_index_writes":           stats.VnodeIndexWrites,
-		"vnode_index_writes_total":     stats.VnodeIndexWritesTotal,
-		"vnode_puts":                   stats.VnodePuts,
-		"vnode_puts_total":             stats.VnodePutsTotal,
-		"read_repairs":                 stats.ReadRepairs,
-		"read_repairs_total":           stats.ReadRepairsTotal,
+		"cluster_fms_active":             stats.ClusterAaeFsmActive,
```

Comment:
I see that the new fields you added follow the pattern the plugin already used. Wouldn't it be better for the plugin not to copy the field names here manually? If they're always going to be the same as the json, you could use golang reflect to read the struct tag and copy out the json name.

Reply:
I was trying to change the bare minimum of the existing implementation but what you suggest makes a lot of sense. I'll figure out a way to avoid duplication and manual work.

Reply:
First things first, thanks for jumping in! 😊

I only started working on this as a quick experiment after realising we were missing lots of important Riak metrics because of them not being supported by this plugin.

At first, the idea was to give it a quick try at adding a few more metrics trying to keep the changes at minimum. However, what you say makes a lot of sense. The current way of adding new metrics is error prone and there is quite a bit of repetition. I will certainly look into how to improve upon it. I'll follow your advice and look into Golang's reflect (Go newbie here).

Reply:
Thanks for the suggestion, @reimda.

I've given it a try [here](https://github.com/influxdata/telegraf/pull/11805/commits/3f9c2f799dc78eb4fdf232e0430762eae8dc8c6a). This is one of the first attempts at writing Go. I hope I didn't do anything too stupid here 🙂.

The main concern I have is the impact this may have on performance given that I believe this code is being executed every time we get a response from the server.

Reply:
Thanks for trying it out. That's the kind of reflect code I was thinking of but you're right to consider performance. I'm not sure if its reflect performance is going to be a problem.

The original pattern could also have performance issues. It decodes into a struct, matching by json key, but then throws away the key. Then when it adds the struct fields to telegraf's metric it hard codes the metric keys which happen to be exactly the same as the json keys. I wouldn't worry about it if there were a small number of fields (say 5-10) but it becomes annoying to deal with in code when there are hundreds.

It might be interesting to compare the performance of both options by writing a couple short go benchmarks (read up on `go test -bench`).

Reply:
There seems to be a quite noticeable difference in performance between the old and new ways of parsing the stats. Here are the results I get when running `go test -bench`.

```
BenchmarkRiak-8      	     961	   1214235 ns/op
BenchmarkRiakOld-8   	   32997	     36223 ns/op
```

This said, one could argue that the new version is fast enough and, perhaps, the gains in developer experience outweigh the loss in performance. @reimda, any thoughts?